iothub/device/src/IotHubBaseClient.cs

```diff
@@ -451,49 +451,42 @@ internal async Task OnMethodCalledAsync(DirectMethodRequest directMethodRequest)
             if (Logging.IsEnabled)
                 Logging.Enter(this, directMethodRequest?.MethodName, directMethodRequest, nameof(OnMethodCalledAsync));
 
-            if (directMethodRequest == null)
+            Debug.Assert(directMethodRequest != null, $"SDK bug where null direct method request is processed.");
+            Debug.Assert(_deviceDefaultMethodCallback != null, $"SDK bug where user callback is null but direct method request is still received.");
+
+            if (directMethodRequest == null
+                || _deviceDefaultMethodCallback == null)
             {
+                if (Logging.IsEnabled)
+                    Logging.Error(
+                        this, 
+                        $"Direct method {directMethodRequest?.RequestId} has request is null '{directMethodRequest == null}' and callback is null '{_deviceDefaultMethodCallback != null}'",
+                        nameof(OnMethodCalledAsync));
                 return;
             }
 
-            DirectMethodResponse directMethodResponse = null;
+            try
+            {
+                DirectMethodResponse directMethodResponse = await _deviceDefaultMethodCallback
+                    .Invoke(directMethodRequest)
+                    .ConfigureAwait(false);
+
+                directMethodResponse.RequestId = directMethodRequest.RequestId;
+                directMethodResponse.PayloadConvention = _clientOptions.PayloadConvention;
 
-            if (_deviceDefaultMethodCallback == null)
+                await SendDirectMethodResponseAsync(directMethodResponse).ConfigureAwait(false);
+            }
+            catch (Exception ex)
             {
-                directMethodResponse = new DirectMethodResponse((int)DirectMethodResponseStatusCode.MethodNotImplemented)
```

Comment:
Don't respond to direct methods on behalf of the user as we don't know how they interpret HTTP status codes. Just log it and don't respond to the request.